docs/adrs/004-privacy-invariants.md

```diff
@@ -0,0 +1,64 @@
+
+1. **PBKDF2 algorithm**: key derivation function with SHA-512 key stretching and 210,000 iterations to hamper brute force, dictionary attacks, and precomputed attacks (rainbow tables) attacks.
+2. **AES-GCM** symmetric encryption scheme, using a random nonce for each encryption operation during wallet initialization.
+3. **SHA-256** hash of the derived key material.
+
+Currently, the key material is stored only in the extension's session storage upon login and is correctly cleared when the extension is locked. This key material is used to unseal the encrypted seed phrase ciphertext stored in local state storage and generate the associated spend key.
+
+#### <ins>Remediations</ins>
+
+While this behavior is correct and expected, there are additional potential privacy-related leaks to consider:
+
+1. The full viewing key material is generated from the spend key, and the plaintext JSON of the full viewing key is stored in local state storage. When the extension is locked, a potential attacker with access to the console can still execute `chrome.storage.local.get().then(console.log)` and access the full viewing key in the clear. To mitigate this, we should:
+
+   - Ensure that all access to sensitive data in state storage is gated behind an authentication check to ensure user has logged in,
+   - Consider storing the encrypted full viewing key or not storing the full viewing key at all and deriving it from the spend key each time it is needed.
+
+2. Require a password for the extension upon mainnet launch and warn users that setting an empty password is not advised.
+3. Avoid transmitting errors by default in case keys are exposed in error data. Currently, `unseal` method for instance has the potential to transmit errors that could expose sensitive key information through the error data. Instead, handle errors securely and generalize the emitted error messages.
+
+Generally, the local and session storage need to be more restricted.
+
+```
+Invariant 2.
+
+The Content Security Policy (CSP) defined in the Manifest V3 file must restrict permissions to only those that are necessary, ensuring no extranous security risks are introduced.
+```
+
+Currently, we define `wasm-unsafe-eval` in the manifest file to enable dynamic loading of WebAssembly modules. Without the `wasm-unsafe-eval` CSP directive, WebAssembly is blocked from loading and executing on the page. However, `wasm-unsafe-eval` opens up the possiblity for attackers to inject and execute malicious WebAssembly code:
+
+```
+const maliciousCode = "(wasm binary code)";
+WebAssembly.instantiate(maliciousCode, {}).then(instance => {
+  // Execute the malicious code
+});
+```
+
+#### <ins>Remediations</ins>
+
+We may instead want to consider serving the Webassembly modules as static files. This approach would allow prax to maintain a more strict content security policy.
```

Comment:
Think the lesson about injection is generally valid, though I don't think we do unsafe eval of any user-submitted data in the extension.

What do you mean about serving wasm as a static file? Wouldn't we always need `wasm-unsafe-eval` given it's a required header to execute wasm in a web context?

Reply:
I was generally pointing out that `wasm-unsafe-eval` is a less strict CSP policy, but I think we need that otherwise I don't think we'd be able to dynamically execute the WASM code.